design-proposals/klip-31-metastore-backups.md

```diff
@@ -0,0 +1,122 @@
+
+i.e. (backup: _confluent-ksql-default__command_topic_1593545289.bak)
+```
+1593545289 {"statement":"CREATE STREAM TEST1(ID INT) WITH(KAFKA_TOPIC='test1', VALUE_FORMAT='JSON');","streamsProperties":{}, ... }
+1593545382 {"statement":"CREATE STREAM TEST2(ID INT) WITH(KAFKA_TOPIC='test2', VALUE_FORMAT='JSON');","streamsProperties":{}, ... }
+1593545529 {"statement":"DROP STREAM TEST1;","streamsProperties":{}, ... }
+```
+
+This format will allow users to easily restore the command_topic using the Kafka producer scripts.
+
+i.e.
+```
+$ cut -d' ' -f2- _confluent-ksql-default__command_topic_1593545289.bak | 
+      kafka-console-producer --broker-list localhost:9092 --topic _confluent-ksql-default__command_topic
+```
+
+### Workflow
+
+When KSQL starts, it always reads the command topic from the beginning. If the backup file does not exist yet, then it will create a new file and start appending each command_topic message to it.
+A backup file is not found if it is the first time backups are enabled, or if the KSQL service.id is different from the ones found in the backup filenames.
+
+If a backup file with the same KSQL service.id exists, then KSQL will append only the new command_topic messages to that file.
+
+KSQL will compare each consumed command_topic message against the next line from the backup file. If both messages are different, then it considers the command_topic as a new restored
+topic, so KSQL will create a new backup file, copy all read messages to the file, and continue appending the new messages to it. If there are no more messages read from the backup file (EOF),
+then KSQL considers this file is the current one and will continue appending messages to it.
+
+KSQL will not use offsets to compare against the backup file. It will assume each offset starts from 1, beginning consuming messages from the topic and reading messages from the file.
+
+### Performance
+
+The KSQL metastore is not expect to grow into a big file. Appending each new command to a file is faster. KSQL command messages are not large to cause slowness during backup file writing.
+It will certainly add extra time during KSQL restarts, but considering the average I/O latency of 10ms per write, then 500 commands would take around 5s to create the initial backup. Also, using
+the same example of 500 commands and an average size of 4k per command (using the ksql_processing_log stream command size as example), then this backup file would grow to 2Mb.
+
+## Test plan
+
+- Unit tests
+- Integration tests to verify backups and restore steps work
+
+## LOEs and Delivery Milestones
+
+It is a small feature that may take a week or less.
+
+## Documentation Updates
+
+Documentation about how to enable backups and restore a command topic manually will be written.
+
+## Compatibility Implications
+
+This is a new feature and does not affect any other feature of KSQL or CP components.
+
+## Security Implications
```

Comment:
@apurvam @derekjn There are some security implications of having a Metastore backup (see below). I was thinking we should not have these backups enabled by default to avoid users leaking data accidentally. But, if the default will be disabled, then users can still at risk of deleting the command_topic and not able to recover it. And if they need a backup option, won't they configure a better backup service in Kafka instead?

Reply:
Sure, we can toggle these backups with config. We can default to off too. But I think the basic premise is that the access to the state store disks also need to be secured: they contain some representation of data in the underlying kafka topics which created those state stores (in a stream table join, the materialized table is basically a copy of the source topic). And those stores don't respect the same ACL rules as the underlying topics anyway: if you have access to them, you can essentially read underlying topic contents. 

So storing the metastore backups along with the state stores is not making the model any worse IMO. Still giving users the option to disable the backup, and also choose the location of the backup makes sense.

Reply:
In simple text, the file grows to 1.7M when using 500 commands. It's not big. I will add gzip compression in the next release among other improvements on formatting and a restore command.